### PR TITLE
Hermes 2 evolution : Closes #4610

### LINKED ISSUE #4610: Hermes 2 evolution
During the discussion in the [development meeting on 2021-04-29](https://indico.cern.ch/event/1020696/) it was decided to evolve hermes2 in a similar fashion to the reaper daemon and finally fully replace hermes1.

For this transition to work the following steps have been identified:

1. Currently the hermes2 daemon is exposed to starvation issue. If one of the destination services is unavailable, the queue cannot be continuously processed and a backlog will rise. To solve this, we propose to expand the messages table with a `service` column and on message creation, directly create duplicates (one for each service) into the table. The hermes2 daemon then processes each service queue independently, thus other queues will not starve if one service is down.
2. Documentation needs to be expanded to properly explain how to setup a monitoring infrastructure based on hermes2.
3. Tests are needed
4. Remove hermes1 and replace by hermes2

### PR DESCRIPTION
**Pull request by cserf**
Recreating PR after closing the first one : https://github.com/rucio/rucio/pull/5127

### CODE REVIEW

File: lib/rucio/core/message.py [hunk 1]
```diff
@@ -43,39 +40,45 @@ def add_message(event_type, payload, session=None):
...
-        raise InvalidObject('Invalid JSON for payload: %(e)s' % locals())
-
-    if len(payload) > 4000:
-        new_message = Message(event_type=event_type, payload='nolimit', payload_nolimit=payload, services=services_list)
-    else:
-        new_message = Message(event_type=event_type, payload=payload, services=services_list)
+    except TypeError as err:  # noqa: F841
+        raise InvalidObject('Invalid JSON for payload: %(err)s' % locals())
+
+    services = services_list.split(',')
```
Comment (joeldierkes) [inline lib/rucio/core/message.py]: We already have a `config_get_int`, `config_get_bool`, ... function in the `common/config.py`. This here is the best opportunity to introduce a `config_get_list`.

File: lib/rucio/core/message.py [hunk 2]
```diff
@@ -43,39 +40,45 @@ def add_message(event_type, payload, session=None):
...
+    for service in services:
+        service.strip(' ')
```
Comment (joeldierkes) [inline lib/rucio/core/message.py]: This should be useless then.

File: lib/rucio/core/message.py [hunk 3]
```diff
@@ -19,17 +19,14 @@
...
-from rucio.common.cache import make_region_memcached
 from rucio.common.config import config_get
 from rucio.common.exception import InvalidObject, RucioException
 from rucio.common.utils import APIEncoder
 from rucio.db.sqla import filter_thread_work
 from rucio.db.sqla.models import Message, MessageHistory
 from rucio.db.sqla.session import transactional_session
 
-REGION = make_region_memcached(expiration_time=900)
+SUPPORTED_SERVICES = ['influx', 'elastic', 'email', 'activemq']
```
Comment (joeldierkes) [inline lib/rucio/core/message.py]: This is a well-defined list, let's use an Enum here. This has several advantages:
- We can test it.
- We don't have to pass strings around, it acts as a "Type"
- We can implement custom function such as `from_string`, which automatically test for corrupted inputs.

File: lib/rucio/core/message.py [hunk 4]
```diff
@@ -43,39 +40,45 @@ def add_message(event_type, payload, session=None):
...
+    for service in services:
+        service.strip(' ')
+        if service not in SUPPORTED_SERVICES:
+            raise RucioException('Unsupported service type (%s)  for messages' % service)
+        if event_type == 'email' and service != 'email':
+            continue
+        if service == 'email' and event_type != 'email':
+            continue
+
+        if len(payload) > 4000:
```
Comment (joeldierkes) [inline lib/rucio/core/message.py]: Hard coded constant.

**Review by joeldierkes — CHANGES_REQUESTED** [review]

**Comment by cserf** [comment]
Last commit is to move to new heartbeat mechanism https://github.com/rucio/rucio/issues/5443

File: lib/rucio/core/message.py (hunk 3 above)
Comment (cserf) [reply lib/rucio/core/message.py]: Enum was implemented. Can you please have a look ?

Comment (joeldierkes) [reply lib/rucio/core/message.py]: :+1:

**Review by joeldierkes — DISMISSED** [review]
lgtm

File: etc/docker/dev/docker-compose.yml [hunk 5]
```diff
@@ -1,13 +1,21 @@
+version: "3"
 services:
   rucio:
     image: docker.io/rucio/rucio-dev
     extra_hosts:
       - "ruciodb:127.0.0.1"
       - "graphite:127.0.0.1"
+      - "influxdb:127.0.0.1"
```
Comment (mlassnig) [inline etc/docker/dev/docker-compose.yml:8]: oof, you want this in the default dev environment?

File: lib/rucio/common/constants.py [hunk 6]
```diff
@@ -59,17 +59,33 @@
...
     AVAILABLE = 'A'
     UNAVAILABLE = 'U'
     COPYING = 'C'
     BEING_DELETED = 'B'
     BAD = 'D'
     TEMPORARY_UNAVAILABLE = 'T'
+
+
+@enum.unique
+class HermesSupportedServices(str, enum.Enum):
```
Comment (mlassnig) [inline lib/rucio/common/constants.py]: Maybe just `HermesService` (singular, and shorter) ?

File: etc/docker/dev/docker-compose.yml (hunk 5 above)
Comment (cserf) [reply etc/docker/dev/docker-compose.yml:8]: Not so big :
```
docker images | grep influx
influxdb                                        latest              2639b070d3b3   4 weeks ago    341MB
```
At least, it's much smaller than Rucio image :sweat_smile:

Comment (mlassnig) [reply etc/docker/dev/docker-compose.yml:8]: ah, that's not what i meant :-) when you start up Elastic&Co it eats away the RAM

File: lib/rucio/common/constants.py (hunk 6 above)
Comment (cserf) [reply lib/rucio/common/constants.py]: Done

File: etc/docker/dev/docker-compose.yml (hunk 5 above)
Comment (bari12) [reply etc/docker/dev/docker-compose.yml:8]: @mlassnig @cserf is this really such an issue? Seems ok to me in terms of mem consumption?

File: lib/rucio/common/config.py [hunk 7]
```diff
@@ -237,6 +237,72 @@ def config_get_bool(section, option, raise_exception=True, default=None, check_c
                 raise err_
 
 
+def config_get_list(section, option, raise_exception=True, default=None, check_config_table=True, session=None,
```
Comment (bari12) [inline lib/rucio/common/config.py:244]: Please rebase this and make it similar to how @joeldierkes adapted the config_get methods. It's much simpler now and should be easy.

**Review by bari12 — COMMENTED** [review]
I think this is GTG, @cserf please just add the one change (config)

File: etc/docker/dev/docker-compose.yml (hunk 5 above)
Comment (cserf) [reply etc/docker/dev/docker-compose.yml:8]: I don't think it's an issue. The tests dont't crash on my machine nor in github

**Review by bari12 — APPROVED** [review]

**Pull request merged**